File: src/fsm.py
class SubitFSM:
    """
    Minimal deterministic FSM for 6‑bit SUBIT states.

    Transitions are defined as:
      { state: { event: next_state } }

    Rules:
      - state: integer 0–63
      - event: string
      - next_state: integer 0–63
      - undefined transitions return the current state
    """

    def __init__(self, transitions: dict):
        self.transitions = transitions

    def next(self, state: int, event: str) -> int:
        return self.transitions.get(state, {}).get(event, state)

File: src/fsm.ts
export class SubitFSM {
  /**
   * Minimal deterministic FSM for 6‑bit SUBIT states.
   *
   * Transitions are defined as:
   *   { state: { event: next_state } }
   *
   * Rules:
   *   - state: number 0–63
   *   - event: string
   *   - next_state: number 0–63
   *   - undefined transitions return the current state
   */
  private transitions: Record<number, Record<string, number>>;

  constructor(transitions: Record<number, Record<string, number>>) {
    this.transitions = transitions;
  }

  next(state: number, event: string): number {
    const row = this.transitions[state];
    if (!row) return state;
    return row[event] ?? state;
  }
}
